app/Http/Controllers/Three/ReactionController.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+namespace Rogue\Http\Controllers\Three;
+
+use Rogue\Models\Post;
+use Rogue\Models\Reaction;
+use Illuminate\Http\Request;
+use Rogue\Http\Controllers\Api\ApiController;
+use Rogue\Http\Transformers\Three\ReactionTransformer;
+
+class ReactionController extends ApiController
+{
+    /**
+     * @var \League\Fractal\TransformerAbstract;
+     */
+    protected $transformer;
+
+    /**
+     * Create a controller instance.
+     *
+     * @return void
+     */
+    public function __construct()
+    {
+        $this->transformer = new ReactionTransformer;
+
+        $this->middleware('auth.api');
+    }
+
+    /**
+     * Store a newly created resource in storage.
+     *
+     * @param \Illuminate\Http\Request $request
+     * @param \Rogue\Models\Post $post
+     * @return \Illuminate\Http\Response
+     */
+    public function store(Request $request, Post $post)
+    {
+        $this->validate($request, [
+            'northstar_id' => 'required|string',
+        ]);
+
+        $userId = $request['northstar_id'];
+        // Check to see if the post has a reaction from this particular user with id of northstar_id.
+        $reaction = Reaction::withTrashed()->where(['northstar_id' => $userId, 'post_id' => $post->id])->first();
+
+        // If a post does not have a reaction from this user, create a reaction.
+        if (is_null($reaction)) {
```

Comment:
This feels like a pretty gnarly set of conditionals, I wonder if we can simplify it at all. Could we extract the `create` below here into a `findOrCreate` and then use `$reaction->wasRecentlyCreated` to determine if we just made it?

```php
$reaction = Reaction::withTrashed()->firstOrCreate(['northstar_id' => $userId, 'post_id' => $post->id]);

if ($reaction->wasRecentlyCreated || $reaction->trashed) {
  // we're adding a new reaction in these cases
} else {
  // otherwise we must be removing
}
```

Reply:
ooo these are so cool @DFurnes ! I never knew about `wasRecentlyCreated` or `firstOrCreate` before! For the conditionals, we would want:
```
if (! $reaction->wasRecentlyCreated || $reaction->trashed()){
```
for that first line though, right? Because if it was just created in line 45, we don't want to create a new reaction? Unless I'm not understanding correctly?

Reply:
Yeah, the logic still gets a little weird since we're soft-deleting these.

My thought was we'd always call `$reaction->untrash()` in that part of the conditional (and it just wouldn't do anything if it's a newly created reaction), since in both these cases we want our end-state (the values for `$code` and `$action`) to be the same.